core/src/main/java/org/apache/iceberg/Snapshots.java

```diff
@@ -0,0 +1,226 @@
+
+    private MicroBatchBuilder(Snapshot snapshot, FileIO io) {
+      this.snapshot = snapshot;
+      this.io = io;
+      this.rowFilter = Expressions.alwaysTrue();
+      this.caseSensitive = true;
+    }
+
+    public MicroBatchBuilder caseSensitive(boolean sensitive) {
+      this.caseSensitive = sensitive;
+      return this;
+    }
+
+    public MicroBatchBuilder filter(Expression newRowFilter) {
+      this.rowFilter = newRowFilter;
+      return this;
+    }
+
+    public MicroBatchBuilder specsById(Map<Integer, PartitionSpec> specs) {
+      this.specsById = specs;
+      return this;
+    }
+
+    public MicroBatch generate(int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      List<ManifestFile> manifests = isStarting ? snapshot.dataManifests() :
+          snapshot.dataManifests().stream().filter(m -> m.snapshotId().equals(snapshot.snapshotId()))
+              .collect(Collectors.toList());
+
+      List<Pair<ManifestFile, Integer>> manifestIndexes = indexManifests(manifests);
+      List<Pair<ManifestFile, Integer>> skippedManifestIndexes = skipManifests(manifestIndexes, startFileIndex);
+
+      return generateMicroBatch(skippedManifestIndexes, startFileIndex, targetSizeInBytes, isStarting);
+    }
+
+    private List<Pair<ManifestFile, Integer>> indexManifests(List<ManifestFile> manifestFiles) {
+      int currentFileIndex = 0;
+      List<Pair<ManifestFile, Integer>> manifestIndexes = Lists.newArrayList();
+
+      for (ManifestFile manifest : manifestFiles) {
+        int filesCount = manifest.addedFilesCount() + manifest.existingFilesCount();
+        manifestIndexes.add(Pair.of(manifest, currentFileIndex));
+        currentFileIndex += filesCount;
+      }
+
+      return manifestIndexes;
+    }
+
+    private List<Pair<ManifestFile, Integer>> skipManifests(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                                            int startFileIndex) {
+      if (startFileIndex == 0) {
+        return indexedManifests;
+      }
+
+      int index = 0;
+      for (Pair<ManifestFile, Integer> manifest : indexedManifests) {
+        if (manifest.second() > startFileIndex) {
+          break;
+        }
+
+        index++;
+      }
+
+      return indexedManifests.subList(index - 1, indexedManifests.size());
+    }
+
+    private MicroBatch generateMicroBatch(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                          int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      if (indexedManifests.isEmpty()) {
+        return new MicroBatch(snapshot.snapshotId(), startFileIndex, startFileIndex + 1, 0L,
+            CloseableIterable.empty(), true);
+      }
+
+      long currentSizeInBytes = 0L;
+      int currentFileIndex = 0;
+      List<CloseableIterable<FileScanTask>> batchTasks = Lists.newArrayList();
+
+      for (Pair<ManifestFile, Integer> pair : indexedManifests) {
+        currentFileIndex = pair.second();
```

Comment:
I think this was a clever idea. I like that this avoids the problem of needing to know the ordinal position of each data file in the manifest. But now the indexes of data files within a manifest are dependent on the row filter.

For a concrete example, consider a table with 4 buckets that is written once an hour with data for all buckets. Each write produces a manifest with 4 data files. So indexed manifests would be `(m1, 0), (m2, 4), (m3, 8)`. If I were to use a filter for just one bucket, `b = 0`, I'd produce files with indexes like this: `(m1-b0.avro, 0), (m2-b0.avro, 4), (m3-b0.avro, 8)`. But I would also get the same indexes if I used a different filter, like `b = 1`: `(m1-b1.avro, 0), (m2-b1.avro, 4), (m3-b1.avro, 8)`.

It's a great way to ensure we can use the index to skip manifests, but I'm a little concerned that making the data file indexes dependent on the data filter might cause problems if that filter is lost or changed. The filter is part of the stream state.

The only thing I can think of where this may be a problem is that you can't change a job and then resume from an old offset safely. If you change the filter, the offset is no longer valid. Do you think we should serialize the filter into offsets to avoid that problem?

I'm interested to hear what @aokolnychyi thinks as well.

Reply:
One more thing to note: it's a little odd at first that this also requires keeping track of the next offset and whether the manifest was exhausted.

Reply:
#1019 could avoid the need for this by giving you the offset within a manifest of each data file.

Reply:
Yes @rdblue , if the filter is changed, the offset stored previously cannot be used again, otherwise it will have a possibility of data loss. I was aware of this issue, but it may not be a big problem for streaming applications, since they're typically long-running and will not change during the runtime.

@aokolnychyi 's proposal above doesn't honor filters, so his proposal doesn't have this problem. But ignoring filters also means to read whole data each time, that would also increase the processing overhead.

Reply:
I think we should go with the approach here, but I would add the filter as a string to the serialized offset. That way, we can check to make sure the strings match and the expression hasn't changed. If the filter is non-null, make sure it matches the current one (string equality would be fine). Once we can project the position column and use that instead of the filter-dependent one, we can omit the filter from future offsets.

Reply:
@rdblue what about ignoring the filters and scan the whole data each time? That way could simplify the current codes.

Reply:
Do you mean to read the manifest without a filter and then apply the filter in this class? I think that would work, then we wouldn't need to keep track of the filter.

Reply:
Changed to ignore the filters when doing `MicroBatch` generating.